File: Text Gbariation/gbariate-text/src/gbariel.cs
using System;
using System.Text;

class Gbariation {

	private enum Direction { UP, DOWN, LEFT, RIGHT }

	private static char[,] keyboard;
	private static Random rand;
	private static readonly int NROW = 3;
	private static readonly int NCOL = 10;
	private static readonly float TYPO_CHANCE = 0.07f;
	private static readonly float SIDE_CHANCE = 0.70f;
	private static readonly float SWAP_CHANCE = 0.06f;
	private static readonly float DOT_CHANCE  = 0.06f;

	static int Main(string[] args) {
		
		bool forever = true;
		rand = new Random();

		char[,] _keyboard = new char[3, 10]{
			{'q', 'w', 'e', 'r', 't', 'y', 'u', 'i', 'o', 'p'},
			{'a', 's', 'd', 'f', 'g', 'h', 'j', 'k', 'l', 'k'},
			{'z', 'x', 'c', 'v', 'b', 'n', 'n', 'm', 'm', ' '}
		};

		keyboard = _keyboard;
		string text = null;

		Console.WriteLine("Keep entering text\n\n");

		try {

			while(forever){
				if(args.Length == 0){
					text = Console.ReadLine();
				} else {
					forever = false;
					for(int i = 0; i < args.Length; i++)
						text += args[i] + " ";
				}

				text = text.ToLower();
				text = GbariateText(new StringBuilder(text));

				Console.WriteLine(text);
			}
		} catch (Exception e){}

		return 0;
	}

	private static string GbariateText(StringBuilder text){

		int i;
		for(i = 0; i < text.Length; i++){
			if(text[i] == ' ' && rand.NextDouble() <= DOT_CHANCE) text[i] = '.';
			else if(rand.NextDouble() <= TYPO_CHANCE) text[i] = Typo(text[i]);
			if(rand.NextDouble() <= SWAP_CHANCE) text = new StringBuilder(Swap(text, i));
		}

		return text.ToString();
	}

	private static char Typo(char c){
		// Horizontal-typo 
		if(rand.NextDouble() <= SIDE_CHANCE){

			if(rand.NextDouble() <= 0.5) return RandomizeChar(c, Direction.RIGHT);
			else return RandomizeChar(c, Direction.LEFT);

		// Vertical-typo
		} else {

			if(rand.NextDouble() <= 0.5) return RandomizeChar(c, Direction.UP);
			else return RandomizeChar(c, Direction.DOWN);
		}
	}

	private static char RandomizeChar(char text, Direction dir){

		int i = 0, j = 0;

		FindChar(text, ref i, ref j);

		switch(dir){
		case Direction.UP:
			if(i > 0) return keyboard[i-1, j];
			break;

		case Direction.DOWN:
			if(i < NROW-1) return keyboard[i+1, j];
			break;

		case Direction.LEFT:
			if(j > 0) return keyboard[i, j-1];
			break;

		case Direction.RIGHT:
			if(j < NCOL-1) return keyboard[i, j+1];
			break;

		default:
			Console.WriteLine("Error");
			return ' ';
		}

		return text;
	}

	private static string Swap(StringBuilder text, int index){
		
		try {
			char aux = text[index];
			text[index] = text[index+1];
			text[index+1] = aux;
			
		} catch (Exception e) {
			return text.ToString();
		}

		return text.ToString();
	}

	private static void FindChar(char text, ref int row, ref int col){
		for(int i = 0; i < NROW; i++){
			for(int j = 0; j < NCOL; j++){
				if(keyboard[i, j] == text){
					row = i;
					col = j;
					return;
				}
			}
		}
	}
}

File: Text Gbariation/gbariate-text/src/gbariel.c
#include <stdio.h>
#include <iostream>
#include <stdlib.h>
#include <string.h>
#include <wchar.h>
#include <time.h>
#include <omp.h>

using namespace std;

#define SIZE 3*1024*1024
#define RAND_MAXF ((float) RAND_MAX)
#define TYPO_CHANCE 0.07f
#define SIDE_CHANCE 0.70f
#define SWAP_CHANCE 0.06f

#define NROW 3
#define NCOL 10

char keyboard[3][10] = {
	{'q', 'w', 'e', 'r', 't', 'y', 'u', 'i', 'o', 'p'},
	{'a', 's', 'd', 'f', 'g', 'h', 'j', 'k', 'l', 'k'},
	{'z', 'x', 'c', 'v', 'b', 'n', 'n', 'm', 'm', ' '}
};

enum Direction { UP, DOWN, LEFT, RIGHT };

void GbariateText(char *text, int n);
char Typo(char text);
char RandomizeChar(char text, enum Direction dir);
void Swap(char *text, int index);
void FindChar(char text, int *i, int *j);
void ToLower(char *text, int n);

int main(int argc, char *argv[]){

	char *text = (char *) malloc(sizeof(char)*SIZE);
	int size;
	
	const wchar_t *str = L"ãéóáö";
	const char *str1 = u8"ãéóáö";
	const char32_t *str2 = U"ãéóáö";

	fwide(stdout, 1);
	wprintf(L"ãéóáö %s\n", str);
	wcout << L"ãéóáö \u0444 " << str << endl;
	printf("ãéóáö %s\n", str1);
	wcout << u8"ãéóáö \u0444 " << str1 << endl;
	printf("ãéóáö %ls\n", str2);
	wcout << U"ãéóáö \u0444 " << str2 << endl;
	return 0;

	srand(time(NULL)*clock());
	text[0] = '\0';

	if(argc == 1){
		printf("Enter text: ");
		scanf("%[^\n]s", text);
	} else for(int i = 1; i < argc; i++){
		text = strcat(text, argv[i]);
		text = strcat(text, " ");
	}

	size = strlen(text);

	ToLower(text, size);
	GbariateText(text, size);

	printf("%s", text);

	free(text);
	return 0;
}

void GbariateText(char *text, int n){

	int i;
	for(i = 0; i < n; i++){
		if(rand()/RAND_MAXF <= TYPO_CHANCE) text[i] = Typo(text[i]);
		if(rand()/RAND_MAXF <= SWAP_CHANCE) Swap(text, i);
	}
}

char Typo(char c){
	// Horizontal-typo 
	if(rand()/RAND_MAXF <= SIDE_CHANCE){

		if(rand()%2) return RandomizeChar(c, RIGHT);
		else return RandomizeChar(c, LEFT);

	// Vertical-typo
	} else {

		if(rand()%2) return RandomizeChar(c, UP);
		else return RandomizeChar(c, DOWN);
	}
}

char RandomizeChar(char text, enum Direction dir){

	int i, j;

	FindChar(text, &i, &j);

	switch(dir){
	case UP:
		if(i > 0) return keyboard[i-1][j];
		break;

	case DOWN:
		if(i < NROW-1) return keyboard[i+1][j];
		break;

	case LEFT:
		if(j > 0) return keyboard[i][j-1];
		break;

	case RIGHT:
		if(j < NCOL-1) return keyboard[i][j+1];
		break;

	default:
		fwprintf(stderr, L"Errõó“ÅGØGX¢çbr\n");
		return -1;
	}

	return text;
}

void Swap(char *text, int index){
	char aux = text[index];
	text[index] = text[index+1];
	text[index+1] = aux;
}

void FindChar(char text, int *row, int *col){

	#pragma omp parallel for
	for(int i = 0; i < NROW; i++){
		for(int j = 0; j < NCOL; j++){
			if(keyboard[i][j] == text){
				*row = i;
				*col = j;
				break;
			}
		}
	}
}

void ToLower(char *text, int n){
	for(int i = 0; i < n; i++){
		if(text[i] >= 'A' && text[i] <= 'Z')
			text[i] += 'a';
	}
}





